delegated_credentials.go

```diff
@@ -0,0 +1,295 @@
+// Copyright 2010 The Go Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+package tls
+
+import (
+	"crypto"
+	"crypto/ecdsa"
+	"crypto/elliptic"
+	"crypto/rand"
+	"crypto/rsa"
+	"crypto/x509"
+	"encoding/asn1"
+	"errors"
+	"fmt"
+	"time"
+)
+
+// After only renewThreshold of credential's validTime is left a new one is created
+const renewThreshold = 0.2
+
+// The Delegation Usage x509 extension identifier - temporary, TBD
+var delegatedCredentialsIdentifier = asn1.ObjectIdentifier{2, 5, 29, 99}
+
+type GetCertificate func(*ClientHelloInfo) (*Certificate, error)
+
+type DelegatedCredential struct {
+	ValidTime int64
+	PublicKey interface{}
+}
+
+type cachedCredential struct {
+	credential []byte
+	validTime  int64
+	privateKey crypto.PrivateKey
+	renewAfter time.Time
+}
+
+type CredentialsConfig struct {
+	Cert      *Certificate
+	Validity  time.Duration
+	scheme    SignatureScheme
+	version   uint16
+	timeFunc  func() time.Time
+	renewChan chan struct{}
+}
+
+var cachedCredentials map[SignatureScheme]*cachedCredential
```

Comment:
This is a package-global variable indexed by the SignatureScheme. This means it will be shared between all instantiations, even those with different certs. You most likely want to keep this local to the CredentialsConfig.